File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/sccp/SccpReplayV1.java
package org.hyperledger.iroha.android.sccp;

import java.io.ByteArrayOutputStream;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import org.hyperledger.iroha.android.model.instructions.TransferWirePayloadEncoder;

/** Canonical SHA-256 sparse-Merkle replay hashing for SCCP final V1. */
public final class SccpReplayV1 {
  public static final int DEPTH = 248;
  private static final byte[] MAGIC = "SCCP-REPLAY-SMT-V1".getBytes(StandardCharsets.US_ASCII);
  private static final BigInteger MAX_U128 = BigInteger.ONE.shiftLeft(128).subtract(BigInteger.ONE);

  /** Closed replay-operation tags. */
  public enum Boundary {
    SORA_OUTBOUND_LOCK(0x01),
    SORA_INBOUND_RELEASE(0x02),
    EVM_SOURCE_BURN(0x10),
    EVM_DESTINATION_MINT(0x11),
    TRON_SOURCE_BURN(0x20),
    TRON_DESTINATION_MINT(0x21),
    TON_BRIDGE_INBOUND_MINT(0x30),
    TON_BRIDGE_OUTBOUND_BURN(0x31),
    TON_MASTER_MINT(0x32),
    TON_MASTER_BURN(0x33),
    TON_WALLET_MINT_CREDIT(0x34),
    TON_WALLET_BURN_DEBIT(0x35),
    TON_WALLET_REFUND_DEBIT(0x36),
    TON_WALLET_REFUND_CREDIT(0x37);

    private final int tag;

    Boundary(final int tag) {
      this.tag = tag;
    }

    public int tag() {
      return tag;
    }
  }

  /** Immutable canonical contract identity. */
  public static final class Actor {
    private final int kind;
    private final byte[] bytes;

    private Actor(final int kind, final byte[] bytes) {
      this.kind = kind;
      this.bytes = bytes.clone();
    }

    public static Actor route() {
      return new Actor(0, new byte[0]);
    }

    public static Actor evm(final byte[] address) {
      return new Actor(1, exact(address, 20, "EVM replay actor", true));
    }

    public static Actor tron(final byte[] address) {
      return new Actor(2, exact(address, 20, "TRON replay actor", true));
    }

    public static Actor ton(final int workchain, final byte[] account) {
      return new Actor(3, concat(i32be(workchain), exact(account, 32, "TON replay actor", true)));
    }
  }

  /** Immutable canonical economic principal. */
  public static final class Principal {
    private final int kind;
    private final byte[] bytes;

    private Principal(final int kind, final byte[] bytes) {
      if (bytes.length == 0 || bytes.length > 0xffff) {
        throw new IllegalArgumentException("invalid replay principal length");
      }
      this.kind = kind;
      this.bytes = bytes.clone();
    }

    /** Construct from the exact canonical Norito AccountId bytes. */
    public static Principal soraAccount(final byte[] canonicalAccountId) {
      return new Principal(0, canonicalSoraAccountId(canonicalAccountId));
    }

    public static Principal evm(final byte[] address) {
      return new Principal(1, exact(address, 20, "EVM replay principal", true));
    }

    public static Principal tron(final byte[] address) {
      return new Principal(2, exact(address, 20, "TRON replay principal", true));
    }

    public static Principal ton(final int workchain, final byte[] account) {
      return new Principal(
          3, concat(i32be(workchain), exact(account, 32, "TON replay principal", true)));
    }
  }

  /** Strict compressed sparse-Merkle witness. */
  public static final class Witness {
    private final byte[] expectedRoot;
    private final byte[] priorDigest;
    private final byte[] bitmap;
    private final List<byte[]> siblings;

    public Witness(
        final byte[] expectedRoot,
        final byte[] priorDigest,
        final byte[] bitmap,
        final List<byte[]> siblings) {
      this.expectedRoot = exact(expectedRoot, 32, "expected shard root", true);
      this.priorDigest = exact(priorDigest, 32, "prior record digest", false);
      this.bitmap = exact(bitmap, 32, "sibling bitmap", false);
      final List<byte[]> copied = new ArrayList<>(siblings.size());
      for (int index = 0; index < siblings.size(); index++) {
        copied.add(exact(siblings.get(index), 32, "sibling", true));
      }
      this.siblings = Collections.unmodifiableList(copied);
    }
  }

  /** Reconstructed witness root and shard. */
  public static final class WitnessRoot {
    private final byte[] root;
    private final byte[] expected;
    private final int shard;

    private WitnessRoot(final byte[] root, final byte[] expected, final int shard) {
      this.root = root.clone();
      this.expected = expected.clone();
      this.shard = shard;
    }

    public byte[] root() {
      return root.clone();
    }

    public byte[] expectedRoot() {
      return expected.clone();
    }

    public boolean matchesExpectedRoot() {
      return Arrays.equals(root, expected);
    }

    public int shard() {
      return shard;
    }
  }

  private SccpReplayV1() {}

  /** Hash one exact production replay domain. */
  public static byte[] domainHash(
      final SccpNetworkV1 source,
      final SccpNetworkV1 target,
      final Boundary boundary,
      final long routeRevision,
      final byte[] routeConfigurationHash,
      final Actor actor) {
    if (!source.isProduction() || !target.isProduction()) {
      throw new IllegalArgumentException("replay domains admit production networks only");
    }
    if (routeRevision <= 0 || routeRevision > 0xffff_ffffL) {
      throw new IllegalArgumentException("route revision must be a nonzero u32");
    }
    if (!validDirection(source, target, boundary, actor.kind)) {
      throw new IllegalArgumentException("invalid replay boundary, direction, or actor");
    }
    return hash(
        MAGIC,
        new byte[] {0},
        unsignedBe(BigInteger.valueOf(source.tag()), 4, "source tag"),
        unsignedBe(BigInteger.valueOf(target.tag()), 4, "target tag"),
        new byte[] {(byte) boundary.tag()},
        unsignedBe(BigInteger.valueOf(routeRevision), 4, "route revision"),
        exact(routeConfigurationHash, 32, "route configuration hash", true),
        new byte[] {(byte) actor.kind},
        unsignedBe(BigInteger.valueOf(actor.bytes.length), 2, "actor length"),
        actor.bytes);
  }

  /** Derive the full replay key; byte zero selects the shard. */
  public static byte[] replayKey(final byte[] domainHash, final byte[] replayId) {
    return hash(
        MAGIC,
        new byte[] {1},
        exact(domainHash, 32, "domain hash", true),
        exact(replayId, 32, "replay id", true));
  }

  /** Hash one occupied leaf record with a canonical scale-9 u128 amount. */
  public static byte[] recordDigest(
      final Boundary operation,
      final byte[] replayId,
      final byte[] payloadSha256,
      final BigInteger amountScale9,
      final Principal principal,
      final byte[] auxiliaryIdentitySha256) {
    if (amountScale9.signum() <= 0 || amountScale9.compareTo(MAX_U128) > 0) {
      throw new IllegalArgumentException("replay amount must be a positive u128");
    }
    final byte[] principalDigest =
        hash(
            MAGIC,
            new byte[] {3, (byte) principal.kind},
            unsignedBe(BigInteger.valueOf(principal.bytes.length), 2, "principal length"),
            principal.bytes);
    final byte[] auxiliary =
        hash(
            MAGIC,
            new byte[] {4, (byte) operation.tag()},
            exact(auxiliaryIdentitySha256, 32, "auxiliary identity SHA-256", true));
    return hash(
        MAGIC,
        new byte[] {2, (byte) operation.tag()},
        exact(replayId, 32, "replay id", true),
        exact(payloadSha256, 32, "payload SHA-256", true),
        unsignedBe(amountScale9, 16, "scale-9 amount"),
        principalDigest,
        auxiliary);
  }

  /** Return all 249 canonical empty hashes in leaf-up order. */
  public static List<byte[]> emptyHashes() {
    final List<byte[]> result = new ArrayList<>(DEPTH + 1);
    result.add(hash(MAGIC, new byte[] {0x10}));
    for (int level = 0; level < DEPTH; level++) {
      result.add(parent(level, result.get(level), result.get(level)));
    }
    final List<byte[]> copied = new ArrayList<>(result.size());
    for (final byte[] value : result) copied.add(value.clone());
    return Collections.unmodifiableList(copied);
  }

  /** Strictly reconstruct one canonical compressed witness. */
  public static WitnessRoot rootFromWitness(
      final byte[] keyValue, final byte[] recordDigest, final Witness witness) {
    final byte[] key = exact(keyValue, 32, "replay key", true);
    if (witness.bitmap[0] != 0) {
      throw new IllegalArgumentException("witness bitmap has reserved high bits");
    }
    int setBits = 0;
    for (final byte item : witness.bitmap) setBits += Integer.bitCount(item & 0xff);
    if (setBits != witness.siblings.size() || setBits > DEPTH) {
      throw new IllegalArgumentException("witness sibling count does not match bitmap");
    }
    final List<byte[]> empty = emptyHashes();
    byte[] current;
    if (recordDigest == null) {
      if (!allZero(witness.priorDigest)) {
        throw new IllegalArgumentException("non-membership witness has an occupied digest");
      }
      current = empty.get(0);
    } else {
      final byte[] digest = exact(recordDigest, 32, "record digest", true);
      if (!Arrays.equals(digest, witness.priorDigest)) {
        throw new IllegalArgumentException("membership witness record digest mismatch");
      }
      current = hash(MAGIC, new byte[] {0x11}, key, digest);
    }
    int supplied = 0;
    for (int level = 0; level < DEPTH; level++) {
      byte[] sibling = empty.get(level);
      if (bit(witness.bitmap, level)) {
        sibling = witness.siblings.get(supplied++);
        if (Arrays.equals(sibling, empty.get(level))) {
          throw new IllegalArgumentException("witness explicitly encodes a default sibling");
        }
      }
      current = bit(key, level) ? parent(level, sibling, current) : parent(level, current, sibling);
    }
    return new WitnessRoot(current, witness.expectedRoot, key[0] & 0xff);
  }

  private static boolean validDirection(
      final SccpNetworkV1 source,
      final SccpNetworkV1 target,
      final Boundary boundary,
      final int actorKind) {
    switch (boundary) {
      case SORA_OUTBOUND_LOCK:
        return source == SccpNetworkV1.SORA_TAIRA && target.isExternal() && actorKind == 0;
      case SORA_INBOUND_RELEASE:
        return source.isExternal() && target == SccpNetworkV1.SORA_TAIRA && actorKind == 0;
      case EVM_SOURCE_BURN:
        return isEvm(source) && target == SccpNetworkV1.SORA_TAIRA && actorKind == 1;
      case EVM_DESTINATION_MINT:
        return source == SccpNetworkV1.SORA_TAIRA && isEvm(target) && actorKind == 1;
      case TRON_SOURCE_BURN:
        return source == SccpNetworkV1.TRON_MAINNET
            && target == SccpNetworkV1.SORA_TAIRA
            && actorKind == 2;
      case TRON_DESTINATION_MINT:
        return source == SccpNetworkV1.SORA_TAIRA
            && target == SccpNetworkV1.TRON_MAINNET
            && actorKind == 2;
      case TON_BRIDGE_INBOUND_MINT:
      case TON_MASTER_MINT:
      case TON_WALLET_MINT_CREDIT:
      case TON_WALLET_REFUND_DEBIT:
      case TON_WALLET_REFUND_CREDIT:
        return source == SccpNetworkV1.SORA_TAIRA
            && target == SccpNetworkV1.TON_MAINNET
            && actorKind == 3;
      case TON_BRIDGE_OUTBOUND_BURN:
      case TON_MASTER_BURN:
      case TON_WALLET_BURN_DEBIT:
        return source == SccpNetworkV1.TON_MAINNET
            && target == SccpNetworkV1.SORA_TAIRA
            && actorKind == 3;
      default:
        return false;
    }
  }

  private static boolean isEvm(final SccpNetworkV1 network) {
    return network == SccpNetworkV1.ETHEREUM_MAINNET || network == SccpNetworkV1.BSC_MAINNET;
  }

  private static byte[] parent(final int level, final byte[] left, final byte[] right) {
    return hash(
        MAGIC,
        new byte[] {0x12},
        unsignedBe(BigInteger.valueOf(level), 2, "tree level"),
        left,
        right);
  }

  private static boolean bit(final byte[] value, final int level) {
    return (value[31 - level / 8] & (1 << (level % 8))) != 0;
  }

  private static byte[] exact(
      final byte[] value, final int length, final String label, final boolean nonzero) {
    if (value == null || value.length != length || (nonzero && allZero(value))) {
      throw new IllegalArgumentException(
          label + " must be " + (nonzero ? "nonzero " : "") + length + " bytes");
    }
    return value.clone();
  }

  private static boolean allZero(final byte[] value) {
    for (final byte item : value) if (item != 0) return false;
    return true;
  }

  private static byte[] canonicalSoraAccountId(final byte[] payload) {
    Objects.requireNonNull(payload, "canonicalAccountId");
    if (payload.length == 0 || payload.length > 0xffff) {
      throw new IllegalArgumentException(
          "SORA replay principal must be canonical nonempty AccountId bytes");
    }
    final String rendered;
    try {
      rendered =
          TransferWirePayloadEncoder.decodeAccountIdPayload(
              payload, SccpV1.TAIRA_I105_DISCRIMINANT_V1);
    } catch (RuntimeException error) {
      throw new IllegalArgumentException(
          "SORA replay principal is not a canonical AccountId", error);
    }
    final byte[] canonical;
    try {
      canonical = TransferWirePayloadEncoder.encodeAccountIdPayload(rendered);
    } catch (RuntimeException error) {
      throw new IllegalArgumentException(
          "SORA replay principal is not a canonical AccountId", error);
    }
    if (!Arrays.equals(canonical, payload)) {
      throw new IllegalArgumentException(
          "SORA replay principal is not the canonical AccountId encoding");
    }
    return payload.clone();
  }

  private static byte[] i32be(final int value) {
    return new byte[] {
      (byte) (value >>> 24), (byte) (value >>> 16), (byte) (value >>> 8), (byte) value
    };
  }

  private static byte[] unsignedBe(
      final BigInteger value, final int width, final String label) {
    if (value.signum() < 0 || value.bitLength() > width * 8) {
      throw new IllegalArgumentException(label + " exceeds u" + (width * 8));
    }
    final byte[] source = value.toByteArray();
    final byte[] result = new byte[width];
    final int count = Math.min(width, source.length);
    System.arraycopy(source, source.length - count, result, width - count, count);
    return result;
  }

  private static byte[] concat(final byte[]... parts) {
    final ByteArrayOutputStream out = new ByteArrayOutputStream();
    for (final byte[] part : parts) out.write(part, 0, part.length);
    return out.toByteArray();
  }

  private static byte[] hash(final byte[]... parts) {
    try {
      final MessageDigest digest = MessageDigest.getInstance("SHA-256");
      for (final byte[] part : parts) digest.update(part);
      return digest.digest();
    } catch (final NoSuchAlgorithmException impossible) {
      throw new AssertionError("JVM lacks mandatory SHA-256", impossible);
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/sccp/SccpReplayV1.kt
package org.hyperledger.iroha.sdk.sccp

import java.io.ByteArrayOutputStream
import java.math.BigInteger
import java.security.MessageDigest
import org.hyperledger.iroha.sdk.core.model.instructions.TransferWirePayloadEncoder

/** Closed replay-boundary tags shared by SORA and destination contracts. */
enum class SccpReplayBoundaryV1(val tag: Int) {
    SORA_OUTBOUND_LOCK(0x01),
    SORA_INBOUND_RELEASE(0x02),
    EVM_SOURCE_BURN(0x10),
    EVM_DESTINATION_MINT(0x11),
    TRON_SOURCE_BURN(0x20),
    TRON_DESTINATION_MINT(0x21),
    TON_BRIDGE_INBOUND_MINT(0x30),
    TON_BRIDGE_OUTBOUND_BURN(0x31),
    TON_MASTER_MINT(0x32),
    TON_MASTER_BURN(0x33),
    TON_WALLET_MINT_CREDIT(0x34),
    TON_WALLET_BURN_DEBIT(0x35),
    TON_WALLET_REFUND_DEBIT(0x36),
    TON_WALLET_REFUND_CREDIT(0x37),
}

/** Canonical contract identity committed by one replay domain. */
class SccpReplayActorV1 private constructor(internal val kind: Int, bytes: ByteArray) {
    internal val bytes = bytes.copyOf()

    companion object {
        @JvmStatic fun route(): SccpReplayActorV1 = SccpReplayActorV1(0, ByteArray(0))

        @JvmStatic fun evm(address: ByteArray): SccpReplayActorV1 =
            SccpReplayActorV1(1, requireExact(address, 20, "EVM replay actor"))

        @JvmStatic fun tron(address: ByteArray): SccpReplayActorV1 =
            SccpReplayActorV1(2, requireExact(address, 20, "TRON replay actor"))

        @JvmStatic fun ton(workchain: Int, account: ByteArray): SccpReplayActorV1 =
            SccpReplayActorV1(
                3,
                signedI32Be(workchain) + requireExact(account, 32, "TON replay actor"),
            )
    }
}

/** Canonical economic principal committed by an occupied replay leaf. */
class SccpReplayPrincipalV1 private constructor(internal val kind: Int, bytes: ByteArray) {
    internal val bytes = bytes.copyOf()

    companion object {
        /** Construct from the exact canonical Norito `AccountId` bytes. */
        @JvmStatic fun soraAccount(canonicalAccountId: ByteArray): SccpReplayPrincipalV1 =
            SccpReplayPrincipalV1(0, canonicalSoraAccountId(canonicalAccountId))

        @JvmStatic fun evm(address: ByteArray): SccpReplayPrincipalV1 =
            SccpReplayPrincipalV1(1, requireExact(address, 20, "EVM replay principal"))

        @JvmStatic fun tron(address: ByteArray): SccpReplayPrincipalV1 =
            SccpReplayPrincipalV1(2, requireExact(address, 20, "TRON replay principal"))

        @JvmStatic fun ton(workchain: Int, account: ByteArray): SccpReplayPrincipalV1 =
            SccpReplayPrincipalV1(
                3,
                signedI32Be(workchain) + requireExact(account, 32, "TON replay principal"),
            )
    }
}

/** Canonically compressed leaf-up sparse-Merkle witness. */
class SccpSparseMerkleWitnessV1(
    expectedShardRoot: ByteArray,
    priorRecordDigest: ByteArray,
    siblingBitmap: ByteArray,
    siblings: List<ByteArray>,
) {
    internal val expectedShardRoot = requireExact(expectedShardRoot, 32, "expected shard root")
    internal val priorRecordDigest = requireExact(
        priorRecordDigest,
        32,
        "prior record digest",
        nonzero = false,
    )
    internal val siblingBitmap = requireExact(
        siblingBitmap,
        32,
        "sibling bitmap",
        nonzero = false,
    )
    internal val siblings = siblings.mapIndexed { index, sibling ->
        requireExact(sibling, 32, "sibling[$index]")
    }
}

/** Result of reconstructing one shard path. */
class SccpReplayWitnessRootV1 internal constructor(
    root: ByteArray,
    expectedRoot: ByteArray,
    val shard: Int,
) {
    private val reconstructed = root.copyOf()
    private val expected = expectedRoot.copyOf()
    val matchesExpectedRoot: Boolean get() = reconstructed.contentEquals(expected)
    fun root(): ByteArray = reconstructed.copyOf()
    fun expectedRoot(): ByteArray = expected.copyOf()
}

/** SHA-256 sparse-Merkle replay hashing shared with Rust and destination runtimes. */
object SccpReplayV1 {
    const val DEPTH = 248
    private val magic = "SCCP-REPLAY-SMT-V1".toByteArray(Charsets.US_ASCII)
    private val maxU128 = BigInteger.ONE.shiftLeft(128).subtract(BigInteger.ONE)

    /** Hash one complete production replay domain. */
    @JvmStatic fun domainHash(
        source: SccpNetworkV1,
        target: SccpNetworkV1,
        boundary: SccpReplayBoundaryV1,
        routeRevision: Long,
        routeConfigurationHash: ByteArray,
        actor: SccpReplayActorV1,
    ): ByteArray {
        require(source.production && target.production) { "replay domains admit production networks only" }
        require(routeRevision in 1..0xffff_ffffL) { "route revision must be a nonzero u32" }
        require(validDirection(source, target, boundary, actor.kind)) {
            "invalid replay boundary, direction, or actor"
        }
        return hash(
            magic,
            byteArrayOf(0),
            unsignedBe(BigInteger.valueOf(source.tag.toLong()), 4, "source tag"),
            unsignedBe(BigInteger.valueOf(target.tag.toLong()), 4, "target tag"),
            byteArrayOf(boundary.tag.toByte()),
            unsignedBe(BigInteger.valueOf(routeRevision), 4, "route revision"),
            requireExact(routeConfigurationHash, 32, "route configuration hash"),
            byteArrayOf(actor.kind.toByte()),
            unsignedBe(BigInteger.valueOf(actor.bytes.size.toLong()), 2, "actor length"),
            actor.bytes,
        )
    }

    /** Derive the full replay key; its first byte selects one of 256 shards. */
    @JvmStatic fun replayKey(domainHash: ByteArray, replayId: ByteArray): ByteArray = hash(
        magic,
        byteArrayOf(1),
        requireExact(domainHash, 32, "domain hash"),
        requireExact(replayId, 32, "replay id"),
    )

    /** Hash one canonical occupied replay record with a scale-9 u128 amount. */
    @JvmStatic fun recordDigest(
        operation: SccpReplayBoundaryV1,
        replayId: ByteArray,
        payloadSha256: ByteArray,
        amountScale9: BigInteger,
        principal: SccpReplayPrincipalV1,
        auxiliaryIdentitySha256: ByteArray,
    ): ByteArray {
        require(amountScale9.signum() > 0 && amountScale9 <= maxU128) {
            "replay amount must be a positive u128"
        }
        val principalDigest = hash(
            magic,
            byteArrayOf(3, principal.kind.toByte()),
            unsignedBe(BigInteger.valueOf(principal.bytes.size.toLong()), 2, "principal length"),
            principal.bytes,
        )
        val auxiliary = hash(
            magic,
            byteArrayOf(4, operation.tag.toByte()),
            requireExact(auxiliaryIdentitySha256, 32, "auxiliary identity SHA-256"),
        )
        return hash(
            magic,
            byteArrayOf(2, operation.tag.toByte()),
            requireExact(replayId, 32, "replay id"),
            requireExact(payloadSha256, 32, "payload SHA-256"),
            unsignedBe(amountScale9, 16, "scale-9 amount"),
            principalDigest,
            auxiliary,
        )
    }

    /** Return all canonical empty hashes in leaf-up order. */
    @JvmStatic fun emptyHashes(): List<ByteArray> {
        val hashes = ArrayList<ByteArray>(DEPTH + 1)
        hashes.add(hash(magic, byteArrayOf(0x10)))
        for (level in 0 until DEPTH) hashes.add(parent(level, hashes[level], hashes[level]))
        return hashes.map { it.copyOf() }
    }

    /** Reconstruct and strictly validate a canonical compressed witness. */
    @JvmStatic fun rootFromWitness(
        keyValue: ByteArray,
        recordDigest: ByteArray?,
        witness: SccpSparseMerkleWitnessV1,
    ): SccpReplayWitnessRootV1 {
        val key = requireExact(keyValue, 32, "replay key")
        require(witness.siblingBitmap[0].toInt() == 0) { "witness bitmap has reserved high bits" }
        val setBits = witness.siblingBitmap.fold(0) { count, byte ->
            count + Integer.bitCount(byte.toInt() and 0xff)
        }
        require(setBits == witness.siblings.size && setBits <= DEPTH) {
            "witness sibling count does not match bitmap"
        }
        val empty = emptyHashes()
        var current = if (recordDigest == null) {
            require(witness.priorRecordDigest.all { it.toInt() == 0 }) {
                "non-membership witness has an occupied digest"
            }
            empty[0]
        } else {
            val digest = requireExact(recordDigest, 32, "record digest")
            require(digest.contentEquals(witness.priorRecordDigest)) {
                "membership witness record digest mismatch"
            }
            hash(magic, byteArrayOf(0x11), key, digest)
        }
        var supplied = 0
        for (level in 0 until DEPTH) {
            val sibling = if (bit(witness.siblingBitmap, level)) {
                witness.siblings[supplied++].also {
                    require(!it.contentEquals(empty[level])) {
                        "witness explicitly encodes a default sibling"
                    }
                }
            } else {
                empty[level]
            }
            current = if (bit(key, level)) parent(level, sibling, current)
            else parent(level, current, sibling)
        }
        return SccpReplayWitnessRootV1(current, witness.expectedShardRoot, key[0].toInt() and 0xff)
    }

    private fun parent(level: Int, left: ByteArray, right: ByteArray): ByteArray = hash(
        magic,
        byteArrayOf(0x12),
        unsignedBe(BigInteger.valueOf(level.toLong()), 2, "tree level"),
        left,
        right,
    )

    private fun bit(value: ByteArray, level: Int): Boolean =
        (value[31 - level / 8].toInt() and (1 shl (level % 8))) != 0

    private fun validDirection(
        source: SccpNetworkV1,
        target: SccpNetworkV1,
        boundary: SccpReplayBoundaryV1,
        actorKind: Int,
    ): Boolean = when (boundary) {
        SccpReplayBoundaryV1.SORA_OUTBOUND_LOCK ->
            source == SccpNetworkV1.SORA_TAIRA && target.isExternal && actorKind == 0
        SccpReplayBoundaryV1.SORA_INBOUND_RELEASE ->
            source.isExternal && target == SccpNetworkV1.SORA_TAIRA && actorKind == 0
        SccpReplayBoundaryV1.EVM_SOURCE_BURN ->
            source in setOf(SccpNetworkV1.ETHEREUM_MAINNET, SccpNetworkV1.BSC_MAINNET) &&
                target == SccpNetworkV1.SORA_TAIRA && actorKind == 1
        SccpReplayBoundaryV1.EVM_DESTINATION_MINT ->
            source == SccpNetworkV1.SORA_TAIRA &&
                target in setOf(SccpNetworkV1.ETHEREUM_MAINNET, SccpNetworkV1.BSC_MAINNET) &&
                actorKind == 1
        SccpReplayBoundaryV1.TRON_SOURCE_BURN ->
            source == SccpNetworkV1.TRON_MAINNET && target == SccpNetworkV1.SORA_TAIRA && actorKind == 2
        SccpReplayBoundaryV1.TRON_DESTINATION_MINT ->
            source == SccpNetworkV1.SORA_TAIRA && target == SccpNetworkV1.TRON_MAINNET && actorKind == 2
        SccpReplayBoundaryV1.TON_BRIDGE_INBOUND_MINT,
        SccpReplayBoundaryV1.TON_MASTER_MINT,
        SccpReplayBoundaryV1.TON_WALLET_MINT_CREDIT,
        SccpReplayBoundaryV1.TON_WALLET_REFUND_DEBIT,
        SccpReplayBoundaryV1.TON_WALLET_REFUND_CREDIT ->
            source == SccpNetworkV1.SORA_TAIRA && target == SccpNetworkV1.TON_MAINNET && actorKind == 3
        SccpReplayBoundaryV1.TON_BRIDGE_OUTBOUND_BURN,
        SccpReplayBoundaryV1.TON_MASTER_BURN,
        SccpReplayBoundaryV1.TON_WALLET_BURN_DEBIT ->
            source == SccpNetworkV1.TON_MAINNET && target == SccpNetworkV1.SORA_TAIRA && actorKind == 3
    }

    private fun hash(vararg parts: ByteArray): ByteArray {
        val digest = MessageDigest.getInstance("SHA-256")
        for (part in parts) digest.update(part)
        return digest.digest()
    }
}

private fun canonicalSoraAccountId(payload: ByteArray): ByteArray {
    require(payload.isNotEmpty() && payload.size <= 0xffff) {
        "SORA replay principal must be canonical nonempty AccountId bytes"
    }
    val rendered = try {
        TransferWirePayloadEncoder.decodeAccountIdPayload(
            payload,
            SccpV1.TAIRA_I105_DISCRIMINANT_V1,
        )
    } catch (error: RuntimeException) {
        throw IllegalArgumentException("SORA replay principal is not a canonical AccountId", error)
    }
    val canonical = try {
        TransferWirePayloadEncoder.encodeAccountIdPayload(rendered)
    } catch (error: RuntimeException) {
        throw IllegalArgumentException("SORA replay principal is not a canonical AccountId", error)
    }
    require(canonical.contentEquals(payload)) {
        "SORA replay principal is not the canonical AccountId encoding"
    }
    return payload.copyOf()
}

private fun requireExact(
    value: ByteArray,
    length: Int,
    label: String,
    nonzero: Boolean = true,
): ByteArray {
    require(value.size == length && (!nonzero || value.any { it.toInt() != 0 })) {
        "$label must be ${if (nonzero) "nonzero " else ""}$length bytes"
    }
    return value.copyOf()
}

private fun signedI32Be(value: Int): ByteArray = byteArrayOf(
    (value ushr 24).toByte(),
    (value ushr 16).toByte(),
    (value ushr 8).toByte(),
    value.toByte(),
)

private fun unsignedBe(value: BigInteger, width: Int, label: String): ByteArray {
    require(value.signum() >= 0 && value.bitLength() <= width * 8) { "$label exceeds u${width * 8}" }
    val source = value.toByteArray()
    val result = ByteArray(width)
    val copyLength = minOf(width, source.size)
    System.arraycopy(source, source.size - copyLength, result, width - copyLength, copyLength)
    return result
}
